src/engine/engine.cpp

```diff
@@ -1670,6 +1678,252 @@ bool DivEngine::addInstrumentFromFile(const char* path) {
         break;
       case DIV_INSFORMAT_BTI:
         break;
+      case DIV_INSFORMAT_S3I:
+        try {
+          reader.seek(0, SEEK_SET);
+
+          uint8_t s3i_type = reader.readC();
+
+          if (s3i_type >= 2) {
+            ins->type = DIV_INS_OPL;
+            // skip internal filename - we'll use the long name description
+            reader.seek(12, SEEK_CUR);
+
+            // skip reserved bytes
+            reader.seek(3, SEEK_CUR);
+
+            // 12-byte opl value
+            uint8_t s3i_Mcharacteristics = reader.readC();
+            uint8_t s3i_Ccharacteristics = reader.readC();
+            uint8_t s3i_Mscaling_output = reader.readC();
+            uint8_t s3i_Cscaling_output = reader.readC();
+            uint8_t s3i_Meg_AD = reader.readC();
+            uint8_t s3i_Ceg_AD = reader.readC();
+            uint8_t s3i_Meg_SR = reader.readC();
+            uint8_t s3i_Ceg_SR = reader.readC();
+            uint8_t s3i_Mwave = reader.readC();
+            uint8_t s3i_Cwave = reader.readC();
+            uint8_t s3i_FeedConnect = reader.readC();
+            
+            DivInstrumentFM::Operator& opM = ins->fm.op[0];
+            DivInstrumentFM::Operator& opC = ins->fm.op[1];
+            ins->fm.ops = 2;
+            opM.mult = s3i_Mcharacteristics & 0xF;
+            opM.ksr = ((s3i_Mcharacteristics >> 4) & 0x1);
+            opM.sus = ((s3i_Mcharacteristics >> 5) & 0x1);
+            opM.vib = ((s3i_Mcharacteristics >> 6) & 0x1);
+            opM.am = ((s3i_Mcharacteristics >> 7) & 0x1);
+            opM.tl = s3i_Mscaling_output & 0x3F;
+            opM.ksl = ((s3i_Mscaling_output >> 6) & 0x3);
+            opM.ar = ((s3i_Meg_AD >> 4) & 0xF);
+            opM.dr = (s3i_Meg_AD & 0xF);
+            opM.rr = (s3i_Meg_SR & 0xF);
+            opM.sl = ((s3i_Meg_SR >> 4) & 0xF);
+            opM.ws = s3i_Mwave;
+
+            ins->fm.alg = (s3i_FeedConnect & 0x1);
+            ins->fm.fb = ((s3i_FeedConnect >> 1) & 0x7);
+
+            opC.mult = s3i_Ccharacteristics & 0xF;
+            opC.ksr = ((s3i_Ccharacteristics >> 4) & 0x1);
+            opC.sus = ((s3i_Ccharacteristics >> 5) & 0x1);
+            opC.vib = ((s3i_Ccharacteristics >> 6) & 0x1);
+            opC.am = ((s3i_Ccharacteristics >> 7) & 0x1);
+            opC.tl = s3i_Cscaling_output & 0x3F;
+            opC.ksl = ((s3i_Cscaling_output >> 6) & 0x3);
+            opC.ar = ((s3i_Ceg_AD >> 4) & 0xF);
+            opC.dr = (s3i_Ceg_AD & 0xF);
+            opC.rr = (s3i_Ceg_SR & 0xF);
+            opC.sl = ((s3i_Ceg_SR >> 4) & 0xF);
+            opC.ws = s3i_Cwave;
+
+            // Skip more stuff we don't need
+            reader.seek(21, SEEK_CUR);
+          } else {
+            logE("S3I PCM samples currently not supported.");
```

Comment:
Reviewer FYI - Stubbed this block in the event S3I PCM sample support may be added in the future.

Reply:
Got it.